Source/Libraries/GSF.Core/Collections/CollectionExtensions.cs

```diff
@@ -422,27 +422,69 @@ public static bool AllParallel<TSource>(this IEnumerable<TSource> source, Func<T
         /// <typeparam name="T"><see cref="Type"/> of elements in the <paramref name="source"/></typeparam>
         /// <param name="source">An enumeration over which to find the median element(s).</param>
         /// <returns>The median item(s) from an enumeration, or <c>null</c> if <paramref name="source"/> is <c>null</c>.</returns>
-        public static T[] Median<T>(this IEnumerable<T> source)
+        public static IEnumerable<T> Median<T>(this IEnumerable<T> source)
         {
             if (source == null)
                 return null;
 
-            T[] values = source as T[] ?? source.ToArray();
+            return source
+                .OrderBy(t => t)
+                .Middle();
+        }
 
-            if (values.Length < 3)
-                return values;
+        /// <summary>
+        /// Gets the middle item(s) from an enumeration, i.e., one return item for odd lengths, two for even lengths.
+        /// </summary>
+        /// <remarks>
+        /// If enumeration length is even, two items will be returned representing the items in the middle of the set;
+        /// otherwise, if the enumeration length is odd, one item will be returned from exactly the middle of the set.
+        /// </remarks>
+        /// <typeparam name="T"><see cref="Type"/> of elements in the <paramref name="source"/></typeparam>
+        /// <param name="source">An enumeration over which to find the middle element(s).</param>
+        /// <returns>The middle item(s) from an enumeration, or <c>null</c> if <paramref name="source"/> is <c>null</c>.</returns>
+        public static IEnumerable<T> Middle<T>(this IEnumerable<T> source)
+        {
+            if (source == null)
+                return null;
 
-            List<T> results = new List<T>();
+            // Instead of source.ToList(), we could use source.Count()
+            // followed by source.Skip(...).Take(...) to avoid memory overhead
+            // but at the cost of enumerating twice
+            if (source is IList<T> list)
+                return MiddleList(list);
+            if (source is ICollection<T> collection)
+                return MiddleCollection(collection);
+            else
+                return MiddleList(source.ToList());
+
+            IEnumerable<T> MiddleList(IList<T> list)
```

Comment:
Not sure on our official sub-function naming conventions - but I tend to make these camel cased (lately anyway)

Reply:
Also `list` hides already declared variable by same name above

Reply:
I suppose the distinction between `IList` and `ICollection` , since `ICollection` has a `Count` property and `IList` implements `ICollection`, is the lack of indexer in `ICollection`.

bah
```

Reply:
I've been using PascalCase for local functions and continuing to use camelCase for delegate variables.

I don't know if there's any useful distinction between a local function and a non-local one when reading the callsite, as opposed to variables where you may need to know if you're modifying class state. By the same token, I'm not sure if there's any useful distinction between local function names and local delegate variables. 🤷